### DISCUSSION TITLE
How to pass secrets variables to your private github packages?

### DISCUSSION THREAD

**Discussion by AdityaVernekar** [community/community#64585] [Packages]
Consider I am having a private npm utility github package that i want to share with my teams. In my package there are some functions dependent on other services via an api key. I m using github actions workflows to deploy my package upgrades. 
How do i set the api keys in my private package. 

Basically I want my teams to use the functions without having to take upon the overhead of passing api keys to the functions. 

Rather some way to embed my secrets in the package and use the functions directly in any other project

**Comment by fortunesteller392** [comment]
To pass secret variables to your private GitHub packages, you can use GitHub Actions and the repository's secrets feature. Here's a step-by-step guide:
1. Create a secret: In your GitHub repository, go to the "Settings" tab, and then click on "Secrets" in the left sidebar. Create a new secret by clicking on the "New repository secret" button. Give your secret a name (e.g., PACKAGE_TOKEN) and enter its value.
2. Configure your workflow: In your repository, create or modify the workflow file (e.g., .github/workflows/main.yml) where you want to use the secret. Add or modify the appropriate step to pass the secret to your GitHub packages.
For example, if you're using the actions/setup-node action to install dependencies and publish a package, you can pass the secret as an environment variable in the job that uses the action:
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v2

      - name: Set up Node.js
        uses: actions/setup-node@v2
        with:
          node-version: '14'

      - name: Install dependencies and publish package
        env:
          PACKAGE_TOKEN: ${{ secrets.PACKAGE_TOKEN }}  # Pass the secret as an environment variable
        run: |
          npm install
          npm publish --registry=https://npm.pkg.github.com/OWNER  # Replace "OWNER" with your GitHub username or organization
In this example, the PACKAGE_TOKEN secret is accessed using the secrets.PACKAGE_TOKEN syntax and passed as an environment variable (PACKAGE_TOKEN) to the npm publish command.
3. Commit and push your changes: Once you've configured your workflow file, commit and push it to your repository.
4. Verify the secret: GitHub will automatically mask the values of secrets in the workflow logs, so you won't see the actual secret value. However, you can verify that the secret is correctly set up by checking the workflow run logs. If there are any issues, you can modify the workflow file and push the changes again.
By following these steps, you'll be able to pass secret variables to your private GitHub packages using GitHub Actions and the repository's secrets feature.

**Comment by fortunesteller392** [comment]
1. Create a Personal Access Token: Generate a Personal Access Token (PAT) in your GitHub account. To do this, go to "Settings" in your GitHub account, navigate to "Developer Settings," and then click on "Personal Access Tokens." Generate a new token with the appropriate permissions (e.g., read:packages, write:packages) and make sure to copy the generated token value.
2. Configure your local environment: In your local development environment or build server, set the PAT as an environment variable. The method for setting environment variables varies depending on your operating system and setup. For example, in Linux/macOS, you can use the export command:
export GITHUB_TOKEN=your_personal_access_token
Alternatively, you can store the token in a file and load it when necessary.
3. Update your package configuration: In your package's configuration file (e.g., package.json for npm packages), update the repository URL to include the PAT. Modify the repository URL by appending :your_personal_access_token to the hostname section. For example:
{
  "name": "your-package",
  "repository": {
    "type": "git",
    "url": "https://github.com/OWNER/REPO.git:your_personal_access_token"
  },
  ...
}
Replace OWNER with your GitHub username or organization and REPO with the name of the repository.
4. Publish your package: Now, when you publish your package using the regular package publishing commands (e.g., npm publish for npm packages), the PAT will be included in the repository URL, granting the necessary access to your private GitHub packages.
npm publish

By following these steps, you can pass secret variables to your private GitHub packages by using a Personal Access Token and updating the package configuration. This approach allows you to control access to your packages without exposing sensitive information directly in your code or configuration files.
